src/rust/engine/fs/src/store.rs

```diff
@@ -256,54 +279,201 @@ mod local {
         .open(path.as_ref())
-        .map_err(|e| format!("Error making env: {}", e.description()))?;
+        .map_err(|e| format!("Error making env: {:?}", e))?;
       let file_database = env
         .create_db(Some("files"), DatabaseFlags::empty())
         .map_err(|e| {
-          format!("Error creating/opening files database: {}", e.description())
+          format!("Error creating/opening files database: {:?}", e)
         })?;
       let directory_database = env
         .create_db(Some("directories"), DatabaseFlags::empty())
         .map_err(|e| {
-          format!(
-            "Error creating/opening directories database: {}",
-            e.description()
-          )
+          format!("Error creating/opening directories database: {:?}", e)
+        })?;
+      let lease_database = env
+        .create_db(Some("leases"), DatabaseFlags::empty())
+        .map_err(|e| {
+          format!("Error creating/opening leases database: {:?}", e)
         })?;
       Ok(ByteStore {
         inner: Arc::new(InnerStore {
-          env: env,
-          pool: pool,
-          file_store: file_database,
-          directory_store: directory_database,
+          env,
+          pool,
+          file_database,
+          directory_database,
+          lease_database,
         }),
       })
     }
-  }
 
-  impl super::ByteStore for ByteStore {
-    fn store_bytes(&self, entry_type: EntryType, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+    pub fn lease_all<Fs: Iterator<Item = Fingerprint>>(
+      &self,
+      fingerprints: Fs,
+    ) -> Result<(), String> {
+      self
+        .inner
+        .env
+        .begin_rw_txn()
+        .map_err(|err| format!("Error making lmdb transaction: {:?}", err))
+        .and_then(|mut txn| {
+          for fingerprint in fingerprints {
+            self.lease(&fingerprint, &mut txn).map_err(|e| {
+              format!("Error leasing fingerprint {}: {:?}", fingerprint, e)
+            })?;
+          }
+          Ok(txn)
+        })
+        .and_then(|txn| {
+          txn.commit().map_err(
+            |e| format!("Error writing lease: {:?}", e),
+          )
+        })
+    }
+
+    fn lease(&self, fingerprint: &Fingerprint, txn: &mut RwTransaction) -> Result<(), lmdb::Error> {
+      let now_since_epoch = time::SystemTime::now()
+        .duration_since(time::UNIX_EPOCH)
+        .expect("Surely you're not before the unix epoch?");
+      let expires_at = (now_since_epoch + time::Duration::from_secs(60 * 60 * 2)).as_secs();
+      let mut buf = [0; 8];
+      LittleEndian::write_u64(&mut buf, expires_at);
+      // This needs to be a standalone variable because otherwise rustfmt will complain.
+      txn.put(
+        self.inner.lease_database,
+        &fingerprint.as_ref(),
+        &buf,
+        WriteFlags::empty(),
+      )
+    }
+
+    ///
+    /// Attempts to shrink the stored files to be no bigger than target_bytes
+    /// (excluding lmdb overhead).
+    ///
```

Comment:
Including a TODO to use the LMDB database statistics when https://github.com/danburkert/lmdb-rs exposes them would be good.

Reply:
Done